cli/src/program.rs

```diff
@@ -557,10 +618,83 @@ pub fn parse_program_subcommand(
                 signers: signer_info.signers,
             }
         }
+        ("upgrade", Some(matches)) => {
+            let sign_only = matches.is_present(SIGN_ONLY_ARG.name);
+            let blockhash_query = BlockhashQuery::new_from_matches(matches);
+
+            let mut bulk_signers = vec![];
+
+            let fee_payer_pubkey = if let Ok((Some(fee_payer_signer), Some(fee_payer_pubkey))) =
+                signer_of(matches, FEE_PAYER_ARG.name, wallet_manager)
+            {
+                bulk_signers.push(Some(fee_payer_signer));
+                fee_payer_pubkey
+            } else {
+                let fee_payer_signer = default_signer.signer_from_path(matches, wallet_manager)?;
+                let fee_payer_pubkey = fee_payer_signer.pubkey();
+                bulk_signers.push(Some(fee_payer_signer));
+                fee_payer_pubkey
+            };
```

Comment:
I think you are correct about `generate_unique_signers`, but:
- it's not very obvious, especially when you read this code (better initialise things related to same entity - `bulk_signers` in this case - is a single place, not spread it through different functions)
- since I'm adding `fee_payer_signer_index` arg to `process_program_deploy`, `process_program_upgrade`, `process_write_buffer` (the purpose being, again, to make it explicit and clearly visible) - we need to get a `fee_payer_pubkey` value here, which `generate_unique_signers` can't help us with

Reply:
Reworked (see latest PR version) in line with approach used for all other commands.